### ISSUE TITLE
Customizing the agent structure output

### ISSUE THREAD

**Issue by mwalidcharrwi** [DLR-RM/stable-baselines3#1849]
### ❓ Question

I am building an RL using sb3 using a custom env and I am creating a custom feature vector and reward, what is the way to verify that the agent is receiving both information and learning based on them?

is there any way to verify the input to agent is being processed and read correctly by the agent?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,

> is there any way to verify the input to agent is being processed and read correctly by the agent?

could you give a bit more context?

if you call `print(model.policy)` you should be able to see the structure of the network used.
For the reward, this is about the environment, not the agent.

**Comment by mwalidcharrwi** [comment]
Hi @araffin, thanks for reaching out.
So the environment observation space is of size 260 (which is a np array) and the reward is returned at the end of the step function that was declared. is there a way for me to verify that the feature vector of size 260 is being red by the agent (my env returns the vector, reward, terminated (bool for end of episode) and info (dictionary) ).

I was trying to print the reward and compute the average of 100 episodes (reset function call) and compare it with the ep_rew_mean and there is a huge difference (although I did check ep_len_mean against my computed mean and they were the same).

Appreciate your valuable guidance and feedback.

**Comment by araffin** [comment]
> I was trying to print the reward and compute the average of 100 episodes (reset function call) and compare it with the ep_rew_mean and there is a huge difference (although I did check ep_len_mean against my computed mean and they were the same).

The wrapper in charge of those statistics is the `Monitor` wrapper: 
https://github.com/DLR-RM/stable-baselines3/blob/a8e905977f3073066eb332f063f6335f355c455a/stable_baselines3/common/monitor.py#L96-L105

you can have a look at the `monitor.csv` file when logging your training (done automatically if you use the RL Zoo).

> the feature vector of size 260 is being red by the agent (my env returns the vector, reward, terminated (bool for end of episode) and info (dictionary) ).

if things run, the agent should use everything that is declared by the env (please use the env checker).
I'm also not sure why you would need a custom features extractor if you only have a vector as input.

**Comment by mwalidcharrwi** [comment]
thanks @araffin for the `Monitor` code snippet.

Now for the reason of using a custom feature is because when I am doing predict on an input to the model, the actions that are being received when `deterministic = True` is getting the actions stuck every single time and when the `deterministic = False` the model is acting more like random and not making use of the training. (for the current time I am training on a single input type to predict on for around 1*10<sup>5</sup> time-steps to illustrate if the model learns and predict on the same input for testing.

I have also placed manual debugging counters to verify the parameters of the model (like ep_len_mean, ep_rew_mean, etc) if they are matching the values, which they are, so it tells me that the training is utilizing the `self.obj` and `reward` however the `model.predict(obs)` is not getting the same behavior as observed in the training.

**Comment by araffin** [comment]
> to illustrate if the model learns and predict on the same input for testing.

Probably similar to https://github.com/DLR-RM/stable-baselines3/issues/1858, anyway, that is what the RL Zoo is for: https://github.com/DLR-RM/rl-baselines3-zoo (making sure training/testing env is exactly the same).

Closing as the original question was answered.

**Comment by mwalidcharrwi** [comment]
Hi @araffin, I checked the referred question. However, I have the seed setup and after training for around 100,000 tilmestep and the `ep_rew_mean ` improving through training but the predict is giving a repetitive action on `deterministic = True`. I have also worked on normalizing the observation space and this outcome still persists.

I saw this issue persisting in several issues and posts and I was wondering if the predict could be stuck in a local minimum which could cause this behavior, and if yes is there any accommodation to overcome that including updating the structure of the observation passed to the agent?
